sim/core/spell_school.go

```diff
@@ -0,0 +1,256 @@
+
+	// Fire x Other
+	SpellSchoolFrostfire,
+	SpellSchoolRadiant,
+	SpellSchoolVolcanic,
+	SpellSchoolShadowflame,
+
+	// Frost x Other
+	SpellSchoolHolyfrost,
+	SpellSchoolFroststorm,
+	SpellSchoolShadowfrost,
+
+	// Holy x Other
+	SpellSchoolHolystorm,
+	SpellSchoolTwilight,
+
+	// Nature x Other
+	SpellSchoolPlague,
+
+	SpellSchoolElemental,
+}
+
+var schoolMaskToIndex = func() map[SpellSchool]stats.SchoolIndex {
+	mti := map[SpellSchool]stats.SchoolIndex{}
+	for i := stats.SchoolIndexNone; i < stats.SchoolLen; i++ {
+		mti[schoolIndexToSchoolMask[i]] = i
+	}
+	return mti
+}()
+
+// LUT for base school indices a (multi)school is made of.
+var schoolIndexToIndices = func() [stats.SchoolLen][]stats.SchoolIndex {
+	arr := [stats.SchoolLen][]stats.SchoolIndex{}
+
+	for schoolIndex := stats.SchoolIndexNone; schoolIndex < stats.SchoolLen; schoolIndex++ {
+		multiMask := SpellSchoolFromIndex(schoolIndex)
+		indexArr := []stats.SchoolIndex{}
+		for baseSchoolIndex := stats.SchoolIndexNone; baseSchoolIndex < stats.PrimarySchoolLen; baseSchoolIndex++ {
+			schoolFlag := SpellSchoolFromIndex(baseSchoolIndex)
+			if multiMask.Matches(schoolFlag) {
+				indexArr = append(indexArr, baseSchoolIndex)
+			}
+		}
+		arr[schoolIndex] = indexArr
+	}
+
+	return arr
+}()
+
+// LUT for resistance stat indices used by each multischool.
+var schoolIndexToResistanceStats = func() [stats.SchoolLen][]stats.Stat {
+	resistances := map[SpellSchool]stats.Stat{
+		SpellSchoolPhysical: stats.Armor, // This is technically physical resistance
+		SpellSchoolArcane:   stats.ArcaneResistance,
+		SpellSchoolFire:     stats.FireResistance,
+		SpellSchoolFrost:    stats.FrostResistance,
+		SpellSchoolNature:   stats.NatureResistance,
+		SpellSchoolShadow:   stats.ShadowResistance,
+	}
+
+	arr := [stats.SchoolLen][]stats.Stat{}
+
+	for schoolIndex := stats.SchoolIndexPhysical; schoolIndex < stats.SchoolLen; schoolIndex++ {
+		schoolMask := SpellSchoolFromIndex(schoolIndex)
+		resiArr := []stats.Stat{}
+		for resiSchool, resiStat := range resistances {
+			if schoolMask.Matches(resiSchool) {
+				resiArr = append(resiArr, resiStat)
+			}
+		}
+		arr[schoolIndex] = resiArr
+	}
+
+	return arr
+}()
+
+// Get spell school mask from school index.
+func SpellSchoolFromIndex(schoolIndex stats.SchoolIndex) SpellSchool {
+	return schoolIndexToSchoolMask[schoolIndex]
+}
+
+func SpellSchoolFromProto(p proto.SpellSchool) SpellSchool {
+	switch p {
+	case proto.SpellSchool_SpellSchoolPhysical:
+		return SpellSchoolPhysical
+	case proto.SpellSchool_SpellSchoolArcane:
+		return SpellSchoolArcane
+	case proto.SpellSchool_SpellSchoolFire:
+		return SpellSchoolFire
+	case proto.SpellSchool_SpellSchoolFrost:
+		return SpellSchoolFrost
+	case proto.SpellSchool_SpellSchoolHoly:
+		return SpellSchoolHoly
+	case proto.SpellSchool_SpellSchoolNature:
+		return SpellSchoolNature
+	case proto.SpellSchool_SpellSchoolShadow:
+		return SpellSchoolShadow
+	default:
+		return SpellSchoolPhysical
```

Comment:
🔨 This feels like one of those cases where physical being the zero value in the proto enum could cause unexpected problems at some point. What do you think about adding a `SpellSchoolUnknown` value as the first entry in that enum?

Reply:
I did not touch this, it's only used for getting the school of the boss auto attacks if I see it right. The proto only has physical=0 to schadow=6, so this function covers all possible cases currently.